orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,172 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.keel.task
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spinnaker.orca.ExecutionStatus
+import com.netflix.spinnaker.orca.KeelService
+import com.netflix.spinnaker.orca.RetryableTask
+import com.netflix.spinnaker.orca.TaskResult
+import com.netflix.spinnaker.orca.igor.ScmService
+import com.netflix.spinnaker.orca.pipeline.model.GitTrigger
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import com.netflix.spinnaker.orca.pipeline.model.Trigger
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Component
+import retrofit.RetrofitError
+import java.util.concurrent.TimeUnit
+
+/**
+ * Task that retrieves a Managed Delivery config manifest from source control via igor, then publishes it to keel,
+ * to support GitOps flows.
+ */
+@Component
+class PublishDeliveryConfigTask
+constructor(
+  private val keelService: KeelService,
+  private val scmService: ScmService,
+  private val objectMapper: ObjectMapper
+) : RetryableTask {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun execute(stage: Stage): TaskResult {
+    val context = objectMapper.convertValue<PublishDeliveryConfigContext>(stage.context)
+    val trigger = stage.execution.trigger
+    val user = trigger.user ?: "anonymous"
+    val manifestLocation = processDeliveryConfigLocation(trigger, context)
+
+    return try {
+      log.debug("Retrieving keel manifest at $manifestLocation")
+      val deliveryConfig = scmService.getDeliveryConfigManifest(
+        context.repoType, context.projectKey, context.repositorySlug, context.directory, context.manifest, context.ref)
+
+      log.debug("Publishing manifest ${context.manifest} to keel on behalf of $user")
+      keelService.publishDeliveryConfig(deliveryConfig, user)
+
+      TaskResult.builder(ExecutionStatus.SUCCEEDED).context(emptyMap<String, Any?>()).build()
+    } catch (e: RetrofitError) {
+      handleFailure(e, context)
+    } catch (e: Exception) {
+      log.error("Unexpected exception while executing {}, aborting.", javaClass.simpleName, e)
+      buildError(e.message)
+    }
+  }
+
+  private fun processDeliveryConfigLocation(trigger: Trigger, context: PublishDeliveryConfigContext): String {
+    if (trigger is GitTrigger) {
```

Comment:
This section basically just looks at the git trigger, if present, and attempts to fill in any blanks in the stage context not filled by the user and that can be inferred from the trigger (namely the `ref`, `repoType`, `projectKey` and `repositorySlug`, all of which are found in the trigger). If a corresponding field is present in the context, the context data wins (hence the `context.field == null` checks).